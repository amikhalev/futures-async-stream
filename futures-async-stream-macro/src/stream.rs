use proc_macro2::{Span, TokenStream};
use quote::{quote, ToTokens};
use syn::{
    parse::{Parse, ParseStream},
    token,
    visit_mut::VisitMut,
    ArgCaptured, Expr, FnArg, FnDecl, Ident, ItemFn, Pat, PatIdent, Result, ReturnType, Token,
    Type, TypeTuple, *,
};

use crate::{
    elision,
    utils::{first_last, respan},
    visitor::{Stream, Visitor},
};

// =================================================================================================
// async_stream

pub(super) fn async_stream(args: TokenStream, input: TokenStream) -> Result<TokenStream> {
    parse_async_stream_fn(args, input)
}

mod kw {
    syn::custom_keyword!(item);
    syn::custom_keyword!(boxed);
    syn::custom_keyword!(boxed_local);
}

// item = <Type>
struct Item(Type);

impl Parse for Item {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let _: kw::item = input.parse()?;
        let _: Token![=] = input.parse()?;
        input.parse().map(Self)
    }
}

struct Args {
    item: Type,
    boxed: ReturnTypeKind,
}

// TODO: rename to `ReturnType`
#[derive(Clone, Copy)]
enum ReturnTypeKind {
    // impl Stream<Item = ..> $(+ $lifetime)?
    Default,
    // Pin<Box<dyn Stream<Item = ..> (+ Send)? $(+ $lifetime)?>>
    Boxed { send: bool },
}

impl Parse for Args {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let mut item = None;
        let mut boxed = ReturnTypeKind::Default;
        while !input.is_empty() {
            if input.peek(kw::boxed) {
                let i: kw::boxed = input.parse()?;
                match boxed {
                    ReturnTypeKind::Default => boxed = ReturnTypeKind::Boxed { send: true },
                    ReturnTypeKind::Boxed { send: true } => {
                        return Err(error!(i, "duplicate `boxed` argument"))
                    }
                    ReturnTypeKind::Boxed { send: false } => {
                        return Err(error!(
                            i,
                            "`boxed` and `boxed_local` cannot be used at the same time."
                        ))
                    }
                }
            } else if input.peek(kw::boxed_local) {
                let i: kw::boxed_local = input.parse()?;
                match boxed {
                    ReturnTypeKind::Default => boxed = ReturnTypeKind::Boxed { send: false },
                    ReturnTypeKind::Boxed { send: false } => {
                        return Err(error!(i, "duplicate `boxed_local` argument"))
                    }
                    ReturnTypeKind::Boxed { send: true } => {
                        return Err(error!(
                            i,
                            "`boxed` and `boxed_local` cannot be used at the same time."
                        ))
                    }
                }
            } else {
                let i: Item = input.parse()?;
                if item.is_some() {
                    return Err(error!(i.0, "duplicate `item` argument"));
                }
                item = Some(i.0);
            }
            let _: Option<Token![,]> = input.parse()?;
        }

        if let Some(item) = item {
            Ok(Self { item, boxed })
        } else {
            let _: kw::item = input.parse()?;
            unreachable!()
        }
    }
}

struct FnSig {
    attrs: Vec<Attribute>,
    vis: Visibility,
    sig: MethodSig,
    block: Block,
    semi: Option<Token![;]>,
}

impl From<ItemFn> for FnSig {
    fn from(item: ItemFn) -> Self {
        Self {
            attrs: item.attrs,
            vis: item.vis,
            sig: MethodSig {
                constness: item.constness,
                asyncness: item.asyncness,
                unsafety: item.unsafety,
                abi: item.abi,
                ident: item.ident,
                decl: *item.decl,
            },
            block: *item.block,
            semi: None,
        }
    }
}

impl From<TraitItemMethod> for FnSig {
    fn from(item: TraitItemMethod) -> Self {
        if let Some(block) = item.default {
            Self { attrs: item.attrs, vis: Visibility::Inherited, sig: item.sig, block, semi: None }
        } else {
            assert!(item.semi_token.is_some());
            Self {
                attrs: item.attrs,
                vis: Visibility::Inherited,
                sig: item.sig,
                block: Block { brace_token: token::Brace::default(), stmts: Vec::new() },
                semi: item.semi_token,
            }
        }
    }
}

fn parse_async_stream_fn(args: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let args: Args = syn::parse2(args)?;
    let item: FnSig = {
        let input2 = input.clone();
        match syn::parse2(input) {
            Ok(item) => ItemFn::into(item),
            Err(e) => match syn::parse2(input2) {
                Ok(item) => TraitItemMethod::into(item),
                Err(_e) => return Err(e),
            },
        }
    };

    if let Some(constness) = item.sig.constness {
        return Err(error!(constness, "async stream may not be const"));
    }
    if let Some(variadic) = item.sig.decl.variadic {
        return Err(error!(variadic, "async stream may not be variadic"));
    }

    if item.sig.asyncness.is_none() {
        return Err(error!(item.sig.decl.fn_token, "async stream must be declared as async"));
    }

    if let ReturnType::Type(_, ty) = &item.sig.decl.output {
        match &**ty {
            Type::Tuple(TypeTuple { elems, .. }) if elems.is_empty() => {}
            _ => return Err(error!(ty, "async stream must return the unit type")),
        }
    }

    Ok(expand_async_stream_fn(item, &args))
}

fn expand_async_stream_fn(item: FnSig, args: &Args) -> TokenStream {
    let FnSig { attrs, vis, sig, mut block, semi } = item;
    let MethodSig { unsafety, abi, ident, decl, .. } = sig;
    let FnDecl { inputs, mut generics, fn_token, .. } = decl;
    let where_clause = &generics.where_clause;

    // Desugar `async fn`
    // from:
    //
    //      #[async_stream(item = u32)]
    //      async fn foo(ref a: u32) {
    //          // ...
    //      }
    //
    // into:
    //
    //      fn foo(__arg_0: u32) -> impl Stream<Item = u32> {
    //          from_generator(static move || {
    //              let ref a = __arg_0;
    //
    //              // ...
    //          })
    //      }
    //
    // We notably skip everything related to `self` which typically doesn't have
    // many patterns with it and just gets captured naturally.
    let mut inputs_no_patterns = Vec::new();
    let mut patterns = Vec::new();
    let mut temp_bindings = Vec::new();
    for (i, input) in inputs.into_iter().enumerate() {
        match input {
            FnArg::Captured(ArgCaptured { pat: Pat::Ident(ref pat), .. })
                if pat.ident == "self" =>
            {
                // `self: Box<Self>` will get captured naturally
                inputs_no_patterns.push(input);
            }
            FnArg::Captured(ArgCaptured {
                pat: pat @ Pat::Ident(PatIdent { by_ref: Some(_), .. }),
                ty,
                colon_token,
            }) => {
                // `ref a: B` (or some similar pattern)
                patterns.push(pat);
                let ident = Ident::new(&format!("__arg_{}", i), Span::call_site());
                temp_bindings.push(ident.clone());
                let pat = PatIdent { by_ref: None, mutability: None, ident, subpat: None }.into();
                inputs_no_patterns.push(ArgCaptured { pat, ty, colon_token }.into());
            }
            _ => {
                // Other arguments get captured naturally
                inputs_no_patterns.push(input);
            }
        }
    }

    // Visit `#[for_await]` and `.await`.
    Visitor::new(Stream).visit_block_mut(&mut block);

    let block_inner = quote! {
        #( let #patterns = #temp_bindings; )*
        #block
    };
    let mut result = TokenStream::new();
    block.brace_token.surround(&mut result, |tokens| {
        block_inner.to_tokens(tokens);
    });
    token::Semi([block.brace_token.span]).to_tokens(&mut result);

    let gen_body_inner = quote! {
        let (): () = #result

        // Ensure that this closure is a generator, even if it doesn't
        // have any `yield` statements.
        #[allow(unreachable_code)]
        {
            return;
            loop { yield ::futures_async_stream::core_reexport::task::Poll::Pending }
        }
    };
    let mut gen_body = TokenStream::new();
    block.brace_token.surround(&mut gen_body, |tokens| {
        gen_body_inner.to_tokens(tokens);
    });

    let item = &args.item;

    // Give the invocation of the `from_generator` function the same span as the `item`
    // as currently errors related to it being a result are targeted here. Not
    // sure if more errors will highlight this function call...
    let output_span = first_last(item);
    let gen_function = quote!(::futures_async_stream::stream::from_generator);
    let gen_function = respan(gen_function, output_span);
    let mut body_inner = quote! {
        #gen_function (static move || -> () #gen_body)
    };
    if let ReturnTypeKind::Boxed { .. } = args.boxed {
        let body = quote! { ::futures_async_stream::alloc_reexport::boxed::Box::pin(#body_inner) };
        body_inner = respan(body, output_span);
    }

    let mut body = TokenStream::new();
    block.brace_token.surround(&mut body, |tokens| {
        body_inner.to_tokens(tokens);
    });

    elision::unelide_lifetimes(&mut generics.params, &mut inputs_no_patterns);
    let lifetimes = generics.lifetimes().map(|l| &l.lifetime);

    let return_ty = match args.boxed {
        ReturnTypeKind::Default => {
            // Raw `impl` breaks syntax highlighting in some editors.
            let impl_token = token::Impl::default();
            quote! {
                #impl_token ::futures_async_stream::stream::Stream<Item = #item> + #(#lifetimes +)*
            }
        }
        ReturnTypeKind::Boxed { send } => {
            let send = if send { Some(quote!(+ Send)) } else { None };
            quote! {
                ::futures_async_stream::core_reexport::pin::Pin<
                    ::futures_async_stream::alloc_reexport::boxed::Box<
                        dyn ::futures_async_stream::stream::Stream<Item = #item> #send + #(#lifetimes +)*
                    >
                >
            }
        }
    };
    let return_ty = respan(return_ty, output_span);

    // FIXME
    let body = semi.map_or_else(|| body, ToTokens::into_token_stream);
    quote! {
        #(#attrs)*
        #vis #unsafety #abi
        #fn_token #ident #generics (#(#inputs_no_patterns),*)
            -> #return_ty
            #where_clause
        #body
    }
}

// =================================================================================================
// async_stream_block

pub(super) fn async_stream_block(input: TokenStream) -> Result<TokenStream> {
    syn::parse2(input).map(expand_async_stream_block)
}

fn expand_async_stream_block(mut expr: Expr) -> TokenStream {
    Visitor::new(Stream).visit_expr_mut(&mut expr);

    let gen_body = quote! {{
        let (): () = #expr;

        // Ensure that this closure is a generator, even if it doesn't
        // have any `yield` statements.
        #[allow(unreachable_code)]
        {
            return;
            loop { yield ::futures_async_stream::core_reexport::task::Poll::Pending }
        }
    }};

    let gen_function = quote!(::futures_async_stream::stream::from_generator);
    quote! {
        #gen_function (static move || -> () #gen_body)
    }
}
