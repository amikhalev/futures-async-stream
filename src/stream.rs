use core::{
    future::Future,
    marker::PhantomData,
    ops::{Generator, GeneratorState},
    pin::Pin,
    task::{Context, Poll},
};
use pin_project::pin_project;
use std::future;

pub use futures_core::stream::Stream;

// =================================================================================================
// GenStream

/// Wrap a generator in a stream.
///
/// This function returns a `GenStream` underneath, but hides it in `impl Trait` to give
/// better error messages (`impl Stream` rather than `GenStream<[closure.....]>`).
#[doc(hidden)]
pub fn from_generator<G, T>(gen: G) -> impl Stream<Item = T>
where
    G: Generator<Yield = Poll<T>, Return = ()>,
{
    GenStream { gen, _phantom: PhantomData }
}

/// A wrapper around generators used to implement `Stream` for `async`/`await` code.
#[pin_project]
#[derive(Debug)]
struct GenStream<G, T> {
    #[pin]
    gen: G,
    _phantom: PhantomData<T>,
}

impl<G, T> Stream for GenStream<G, T>
where
    G: Generator<Yield = Poll<T>, Return = ()>,
{
    type Item = T;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let this = self.project();
        future::set_task_context(cx, || match this.gen.resume() {
            GeneratorState::Yielded(x) => x.map(Some),
            GeneratorState::Complete(()) => Poll::Ready(None),
        })
    }
}

// =================================================================================================
// Poll

/// Polls a stream in the current thread-local task waker.
#[doc(hidden)]
pub fn poll_next_with_tls_context<S>(s: Pin<&mut S>) -> Poll<Option<S::Item>>
where
    S: Stream,
{
    future::get_task_context(|cx| S::poll_next(s, cx))
}

// This is equivalent to the `std::future::poll_with_tls_context` method.
// The `.await` called in `async_stream` needs to be adjusted to yield `Poll`,
// but for this purpose, we don't want the user to use `#![feature(gen_future)]`.
/// Polls a future in the current thread-local task waker.
#[doc(hidden)]
pub fn poll_with_tls_context<F>(f: Pin<&mut F>) -> Poll<F::Output>
where
    F: Future,
{
    future::get_task_context(|cx| F::poll(f, cx))
}

// =================================================================================================
// Next

// This is equivalent to the `futures::StreamExt::next` method.
// But we want to make this crate dependency as small as possible, so we define our `next` function.
#[doc(hidden)]
pub fn next<S>(stream: &mut S) -> impl Future<Output = Option<S::Item>> + '_
where
    S: Stream + Unpin,
{
    Next { stream }
}

#[derive(Debug)]
struct Next<'a, S> {
    stream: &'a mut S,
}

impl<S> Unpin for Next<'_, S> where S: Unpin {}

impl<S> Future for Next<'_, S>
where
    S: Stream + Unpin,
{
    type Output = Option<S::Item>;

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        Pin::new(&mut self.stream).poll_next(cx)
    }
}
