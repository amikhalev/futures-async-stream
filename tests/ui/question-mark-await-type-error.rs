// compile-fail

#![deny(warnings)]
#![feature(generators, stmt_expr_attributes, proc_macro_hygiene)]

use futures::stream;
use futures_async_stream::for_await;

async fn async_fn() {
    for _i in 1..2 {
        async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    }
}

#[async_stream(item = i32)]
async fn async_stream_fn() {
    for _i in 1..2 {
        async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    }
}

// FIXME: this is a compiler bug and probably the fix will require https://github.com/rust-lang/rust/issues/43081
async fn async_fn_and_for_await() {
    #[for_await]
    //~^ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    for _i in stream::iter(1..2) {
        async {}.await?;
    }
}

// FIXME: this is a compiler bug and probably the fix will require https://github.com/rust-lang/rust/issues/43081
#[async_stream(item = i32)]
async fn async_stream_fn_and_for_await() {
    #[for_await]
    //~^ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    for _i in stream::iter(1..2) {
        async {}.await?;
    }
}

fn main() {}
