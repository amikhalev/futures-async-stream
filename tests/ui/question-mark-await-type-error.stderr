error[E0277]: the `?` operator can only be applied to values that implement `std::ops::Try`
  --> $DIR/question-mark-await-type-error.rs:16:9
   |
16 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |         ^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `()`
   |
   = help: the trait `std::ops::Try` is not implemented for `()`
   = note: required by `std::ops::Try::into_result`

error[E0698]: type inside `async` object must be known in this context
  --> $DIR/question-mark-await-type-error.rs:15:15
   |
15 |     for _i in 1..2 {
   |               ^^^^ cannot infer type for `{integer}`
   |
note: the type is part of the `async` object because of this `await`
  --> $DIR/question-mark-await-type-error.rs:16:9
   |
16 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |         ^^^^^^^^^^^^^^

error[E0698]: type inside `async` object must be known in this context
  --> $DIR/question-mark-await-type-error.rs:15:15
   |
15 |     for _i in 1..2 {
   |               ^ cannot infer type for `{integer}`
   |
note: the type is part of the `async` object because of this `await`
  --> $DIR/question-mark-await-type-error.rs:16:9
   |
16 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |         ^^^^^^^^^^^^^^

error[E0698]: type inside `async` object must be known in this context
  --> $DIR/question-mark-await-type-error.rs:15:18
   |
15 |     for _i in 1..2 {
   |                  ^ cannot infer type for `{integer}`
   |
note: the type is part of the `async` object because of this `await`
  --> $DIR/question-mark-await-type-error.rs:16:9
   |
16 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |         ^^^^^^^^^^^^^^

error[E0698]: type inside `async` object must be known in this context
  --> $DIR/question-mark-await-type-error.rs:15:15
   |
15 |     for _i in 1..2 {
   |               ^^^^ cannot infer type for `{integer}`
   |
note: the type is part of the `async` object because of this `await`
  --> $DIR/question-mark-await-type-error.rs:16:9
   |
16 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |         ^^^^^^^^^^^^^^

error[E0698]: type inside `async` object must be known in this context
  --> $DIR/question-mark-await-type-error.rs:15:9
   |
15 |     for _i in 1..2 {
   |         ^^ cannot infer type for `{integer}`
   |
note: the type is part of the `async` object because of this `await`
  --> $DIR/question-mark-await-type-error.rs:16:9
   |
16 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |         ^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be applied to values that implement `std::ops::Try`
  --> $DIR/question-mark-await-type-error.rs:29:5
   |
29 |     #[for_await]
   |     ^^^^^^^^^^^^ the `?` operator cannot be applied to type `()`
   |
   = help: the trait `std::ops::Try` is not implemented for `()`
   = note: required by `std::ops::Try::into_result`

error[E0277]: the `?` operator can only be applied to values that implement `std::ops::Try`
  --> $DIR/question-mark-await-type-error.rs:23:18
   |
23 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |                  ^^^^^^ the `?` operator cannot be applied to type `()`
   |
   = help: the trait `std::ops::Try` is not implemented for `()`
   = note: required by `std::ops::Try::into_result`

error[E0698]: type inside generator must be known in this context
  --> $DIR/question-mark-await-type-error.rs:22:15
   |
22 |     for _i in 1..2 {
   |               ^^^^ cannot infer type for `{integer}`
   |
note: the type is part of the generator because of this `yield`
  --> $DIR/question-mark-await-type-error.rs:23:18
   |
23 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |                  ^^^^^

error[E0698]: type inside generator must be known in this context
  --> $DIR/question-mark-await-type-error.rs:22:15
   |
22 |     for _i in 1..2 {
   |               ^ cannot infer type for `{integer}`
   |
note: the type is part of the generator because of this `yield`
  --> $DIR/question-mark-await-type-error.rs:23:18
   |
23 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |                  ^^^^^

error[E0698]: type inside generator must be known in this context
  --> $DIR/question-mark-await-type-error.rs:22:18
   |
22 |     for _i in 1..2 {
   |                  ^ cannot infer type for `{integer}`
   |
note: the type is part of the generator because of this `yield`
  --> $DIR/question-mark-await-type-error.rs:23:18
   |
23 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |                  ^^^^^

error[E0698]: type inside generator must be known in this context
  --> $DIR/question-mark-await-type-error.rs:22:15
   |
22 |     for _i in 1..2 {
   |               ^^^^ cannot infer type for `{integer}`
   |
note: the type is part of the generator because of this `yield`
  --> $DIR/question-mark-await-type-error.rs:23:18
   |
23 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |                  ^^^^^

error[E0698]: type inside generator must be known in this context
  --> $DIR/question-mark-await-type-error.rs:22:9
   |
22 |     for _i in 1..2 {
   |         ^^ cannot infer type for `{integer}`
   |
note: the type is part of the generator because of this `yield`
  --> $DIR/question-mark-await-type-error.rs:23:18
   |
23 |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
   |                  ^^^^^

error[E0277]: the `?` operator can only be applied to values that implement `std::ops::Try`
  --> $DIR/question-mark-await-type-error.rs:42:18
   |
42 |         async {}.await?;
   |                  ^^^^^^ the `?` operator cannot be applied to type `()`
   |
   = help: the trait `std::ops::Try` is not implemented for `()`
   = note: required by `std::ops::Try::into_result`

Some errors have detailed explanations: E0277, E0698.
For more information about an error, try `rustc --explain E0277`.
